### PR TITLE
[Packager] Fix @providesModule not being ignored properly

### PR DESCRIPTION
**Pull request by skevy**
There's quite a bit of code scattered around the packager regarding ignoring the `@providesModule` Haste pragma in any file that isn't in `react-native`, `react-tools` or `parse`. There is even a (passing) test case.

However, there's an edge case.

Take, for example, `fbjs`. It has a module inside of it called `ErrorUtils`. `react-relay` requires this file normally, in Common.JS style, by doing `require('fbjs/libs/ErrorUtils')`. But when `react-native` attempts to require `ErrorUtils` using the HasteModule format (in it's JavaScript initialization), it resolves the `fbjs` `ErrorUtils` module, instead of RN's `ErrorUtils`.

This happens, it turns out, because when a module is read (in `Module._read`), it's not caring about whether or not it should pay attention to `@providesModule`, and is just assigning the `@providesModule` value as the id of the module no matter what. Then when `Module.getName` is called, it will always use that `data.id` that was set, thus creating the wrong dependency tree.

This PR fixes that. :)

### CODE REVIEW

File: packager/react-packager/src/DependencyResolver/__tests__/Module-test.js [hunk 1]
```diff
@@ -22,6 +22,7 @@ jest
 var Fastfs = require('../fastfs');
 var Module = require('../Module');
 var ModuleCache = require('../ModuleCache');
+var Helpers = require('../DependencyGraph/Helpers');
```
Comment (ide) [inline packager/react-packager/src/DependencyResolver/__tests__/Module-test.js]: We might want to rename this module to be more descriptive (not sure what though. DependencyGraphHelpers?) since it's being used outside of DependencyGraph now.

Comment (skevy) [inline packager/react-packager/src/DependencyResolver/__tests__/Module-test.js]: Totally agree.

**Comment by skevy** [comment]
@martinbigio should I update and rebase this PR, or are you guys fixing this issue in a different way?

**Comment by vjeux** [comment]
@skevy fyi Martin is in PTO for three weeks starting today :x Let me take it

**Comment by skevy** [comment]
Yay time off! I wish I could get some of that. :(

**Comment by skevy** [comment]
Thanks for taking over!

**Comment by vjeux** [comment]
@facebook-github-bot import

File: packager/react-packager/src/DependencyResolver/Module.js [hunk 2]
```diff
@@ -85,7 +86,7 @@ class Module {
       this._reading = this._fastfs.readFile(this.path).then(content => {
         const data = {};
         const moduleDocBlock = docblock.parseAsObject(content);
-        if (moduleDocBlock.providesModule || moduleDocBlock.provides) {
+        if (!this._helpers.isNodeModulesDir(this.path) && (moduleDocBlock.providesModule || moduleDocBlock.provides)) {
```
Comment (vjeux) [inline packager/react-packager/src/DependencyResolver/Module.js]: nit: 80 columns.

```javascript
if (!this._helpers.isNodeModulesDir(this.path) &&
    (moduleDocBlock.providesModule || moduleDocBlock.provides)) {
```

File: packager/react-packager/src/DependencyResolver/__tests__/Module-test.js [hunk 3]
```diff
@@ -39,7 +40,7 @@ describe('Module', () => {
       );
 
       return fastfs.build().then(() => {
-        var module = new Module('/root/index.js', fastfs, new ModuleCache(fastfs));
+        var module = new Module('/root/index.js', fastfs, new ModuleCache(fastfs), null, new Helpers());
```
Comment (vjeux) [inline packager/react-packager/src/DependencyResolver/__tests__/Module-test.js]: nit: 80 columns

```javascript
var module = new Module(
  '/root/index.js',
  fastfs,
  new ModuleCache(fastfs),
  null, // whatever-the-name-of-the-arg-is
  new Helpers()
);

Comment (vjeux) [inline packager/react-packager/src/DependencyResolver/__tests__/Module-test.js]: add an inline comment for the `null` argument

**Comment by vjeux** [comment]
@facebook-github-bot import

**Comment by vjeux** [comment]
```
   Error  (FBESLINT0) ESLint reported an error.
    "fileWatcher" is not defined. (no-undef)

            2325 
            2326       var dgraph = new DependencyGraph({
            2327         roots: [root],
    >>>     2328         fileWatcher: fileWatcher,
            2329         assetExts: ['png', 'jpg'],
            2330         cache: cache,
            2331       });

   Error  (FBESLINT0) ESLint reported an error.
    "cache" is not defined. (no-undef)

            2327         roots: [root],
            2328         fileWatcher: fileWatcher,
            2329         assetExts: ['png', 'jpg'],
    >>>     2330         cache: cache,
            2331       });
            2332       return getOrderedDependenciesAsJSON(dgraph, '/root/index.js').then(function(deps) {
            2333         expect(deps)
```

The tests are not passing because it uses variables that are never defined :x

**Comment by vjeux** [comment]
Could you rebase? :) I have no idea what i'm doing here :x

**Comment by skevy** [comment]
Sorry @vjeux, I didn't mean to push. I'm working on this now.

**Comment by skevy** [comment]
This issue with rebasing definitely has something to do with the work @cpojer is doing on the packager I think. I need to take a little bit of a look at what he's up to, and complete the rebase tomorrow when I'm fresh. Thanks for being johnny on the spot with this @vjeux...hopefully we can get it all fixed up and ready to ship tomorrow! :)

**Comment by cpojer** [comment]
I'm about to pull out the resolver into a separate repo sometime this week. It would be great if you could rebase this asap and land it, so I don't run into issues with syncing this. Sorry for the trouble!

File: packager/react-packager/src/DependencyResolver/Module.js [hunk 4]
```diff
@@ -15,7 +15,7 @@ const replacePatterns = require('./replacePatterns');
 
 class Module {
 
-  constructor(file, fastfs, moduleCache, cache) {
+  constructor(file, fastfs, moduleCache, cache, helpers) {
```
Comment (cpojer) [inline packager/react-packager/src/DependencyResolver/Module.js]: we should probably change this into taking a single object, and use destructuring in here.

File: packager/react-packager/src/DependencyResolver/DependencyGraph/DependencyGraphHelpers.js [hunk 5]
```diff
@@ -10,7 +10,7 @@
 
 const path = require('path');
 
-class Helpers {
+class DependencyGraphHelpers {
```
Comment (cpojer) [inline packager/react-packager/src/DependencyResolver/DependencyGraph/DependencyGraphHelpers.js]: :+1:

File: packager/react-packager/src/DependencyResolver/Module.js (hunk 2 above)
Comment (cpojer) [inline packager/react-packager/src/DependencyResolver/Module.js]: Personally I prefer:

```js
if (!this._helpers.isNodeModulesDir(this.path) &&
    (moduleDocBlock.providesModule || moduleDocBlock.provides)
) {
```

but I don't wanna bikeshed :P

**Comment by cpojer** [comment]
Also cc @davidaurelio who is working on packager as well.

**Comment by skevy** [comment]
@cpojer is tomorrow morning ok?

**Comment by cpojer** [comment]
sure, don't stress yourself out :)

File: packager/react-packager/src/DependencyResolver/Module.js [hunk 6]
```diff
@@ -15,7 +15,7 @@ const extractRequires = require('./lib/extractRequires');
 
 class Module {
 
-  constructor(file, fastfs, moduleCache, cache, extractor) {
+  constructor({ file, fastfs, moduleCache, cache, extractor, depGraphHelpers }) {
```
Comment (skevy) [inline packager/react-packager/src/DependencyResolver/Module.js]: @cpojer followed your recco and added destructuring here

**Comment by skevy** [comment]
@cpojer @vjeux  finally got this rebased and updated. Sorry for the wait!

**Comment by vjeux** [comment]
@facebook-github-bot shipit

File: packager/react-packager/src/DependencyResolver/Module.js (hunk 6 above)
Comment (cpojer) [inline packager/react-packager/src/DependencyResolver/Module.js]: sweeet!

**Comment by skevy** [comment]
@vjeux @cpojer just wanted to check on the status of this. Do you need me to do anything else here? Would be handy to get this fixed.

**Comment by vjeux** [comment]
Sorry about that, I thought it landed. There's a few jest errors. Going to take a look at them tomorrow, but maybe you have an idea

```
Running 163 test suites...
 FAIL  react-native-github/packager/react-packager/src/DependencyResolver/DependencyGraph/__tests__/DependencyGraph-test.js (12.93s)
● DependencyGraph › get sync dependencies › it should get dependencies with relative assets
  - TypeError: Cannot read property 'isNodeModulesDir' of undefined
        at react-native-github/packager/react-packager/src/DependencyResolver/Module.js:88:35
        at tryCallOne (react-native-github/node_modules/promise/lib/core.js:37:12)
        at react-native-github/node_modules/promise/lib/core.js:103:15
        at flush (react-native-github/node_modules/promise/node_modules/asap/raw.js:50:29)
        at react-native-github/node_modules/jest-cli/src/lib/FakeTimers.js:355:18
        at doNTCallback0 (../node.js:407:9)
        at process._tickCallback (../node.js:336:13)
● DependencyGraph › get sync dependencies › it should get dependencies with assets and resolution
  - TypeError: Cannot read property 'isNodeModulesDir' of undefined
        at react-native-github/packager/react-packager/src/DependencyResolver/Module.js:88:35
        at tryCallOne (react-native-github/node_modules/promise/lib/core.js:37:12)
        at react-native-github/node_modules/promise/lib/core.js:103:15
        at flush (react-native-github/node_modules/promise/node_modules/asap/raw.js:50:29)
        at react-native-github/node_modules/jest-cli/src/lib/FakeTimers.js:355:18
        at doNTCallback0 (../node.js:407:9)
        at process._tickCallback (../node.js:336:13)
● DependencyGraph › get sync dependencies › it should respect platform extension in assets
  - TypeError: Cannot read property 'isNodeModulesDir' of undefined
        at react-native-github/packager/react-packager/src/DependencyResolver/Module.js:88:35
        at tryCallOne (react-native-github/node_modules/promise/lib/core.js:37:12)
        at react-native-github/node_modules/promise/lib/core.js:103:15
        at flush (react-native-github/node_modules/promise/node_modules/asap/raw.js:50:29)
        at react-native-github/node_modules/jest-cli/src/lib/FakeTimers.js:355:18
        at doNTCallback0 (../node.js:407:9)
        at process._tickCallback (../node.js:336:13)
● DependencyGraph › get sync dependencies › it Deprecated and relative assets can live together
  - TypeError: Cannot read property 'isNodeModulesDir' of undefined
        at react-native-github/packager/react-packager/src/DependencyResolver/Module.js:88:35
        at tryCallOne (react-native-github/node_modules/promise/lib/core.js:37:12)
        at react-native-github/node_modules/promise/lib/core.js:103:15
        at flush (react-native-github/node_modules/promise/node_modules/asap/raw.js:50:29)
        at react-native-github/node_modules/jest-cli/src/lib/FakeTimers.js:355:18
        at doNTCallback0 (../node.js:407:9)
        at process._tickCallback (../node.js:336:13)
● DependencyGraph › file watch updating › it updates module dependencies on relative asset add
  - TypeError: Cannot read property 'isNodeModulesDir' of undefined
        at react-native-github/packager/react-packager/src/DependencyResolver/Module.js:88:35
        at tryCallOne (react-native-github/node_modules/promise/lib/core.js:37:12)
        at react-native-github/node_modules/promise/lib/core.js:103:15
        at flush (react-native-github/node_modules/promise/node_modules/asap/raw.js:50:29)
        at react-native-github/node_modules/jest-cli/src/lib/FakeTimers.js:355:18
        at doNTCallback0 (../node.js:407:9)
        at process._tickCallback (../node.js:336:13)
 FAIL  react-native-github/packager/react-packager/src/BundlesLayout/__tests__/BundlesLayoutIntegration-test.js (7.691s)
● BundlesLayout › generate › it should put image dependencies into separate bundles
  - TypeError: Cannot read property 'isNodeModulesDir' of undefined
        at react-native-github/packager/react-packager/src/DependencyResolver/Module.js:88:35
        at tryCallOne (react-native-github/node_modules/promise/lib/core.js:37:12)
        at react-native-github/node_modules/promise/lib/core.js:103:15
        at flush (react-native-github/node_modules/promise/node_modules/asap/raw.js:50:29)
        at react-native-github/node_modules/jest-cli/src/lib/FakeTimers.js:355:18
        at doNTCallback0 (../node.js:407:9)
        at process._tickCallback (../node.js:336:13)
● BundlesLayout › generate › it should put image dependencies across bundles
  - TypeError: Cannot read property 'isNodeModulesDir' of undefined
        at react-native-github/packager/react-packager/src/DependencyResolver/Module.js:88:35
        at tryCallOne (react-native-github/node_modules/promise/lib/core.js:37:12)
        at react-native-github/node_modules/promise/lib/core.js:103:15
        at flush (react-native-github/node_modules/promise/node_modules/asap/raw.js:50:29)
        at react-native-github/node_modules/jest-cli/src/lib/FakeTimers.js:355:18
        at doNTCallback0 (../node.js:407:9)
        at process._tickCallback (../node.js:336:13)
● BundlesLayout › generate › it could async require asset
  - TypeError: Cannot read property 'isNodeModulesDir' of undefined
        at react-native-github/packager/react-packager/src/DependencyResolver/Module.js:88:35
        at tryCallOne (react-native-github/node_modules/promise/lib/core.js:37:12)
        at react-native-github/node_modules/promise/lib/core.js:103:15
        at flush (react-native-github/node_modules/promise/node_modules/asap/raw.js:50:29)
        at react-native-github/node_modules/jest-cli/src/lib/FakeTimers.js:355:18
        at doNTCallback0 (../node.js:407:9)
        at process._tickCallback (../node.js:336:13)
Running 145 test suites...Error: 
 stack: Error: Requiring unknown module "InitializeJavaScriptAppEngine". If you are sure the module is there, try restarting the packager.
    at requireImpl (evalmachine.<anonymous>:48:7)
    at requireImpl (evalmachine.<anonymous>:34:27)
    at require (evalmachine.<anonymous>:26:8)
    at evalmachine.<anonymous>:6120:2
    at ContextifyScript.Script.runInNewContext (vm.js:18:15)
    at Object.exports.runInNewContext (vm.js:49:17)
    at /data/sandcastle/boxes/instance-fbobjc-fbsource-catalyst-jest-macosx-local/fbobjc/Libraries/FBReactKit/js/RKJSModules/CLIIntegrationTest/__tests__/CLIIntegration-test.js:128:15
    at tryCallOne (/data/sandcastle/boxes/instance-fbobjc-fbsource-catalyst-jest-macosx-local/fbobjc/Libraries/FBReactKit/js/react-native-github/node_modules/promise/lib/core.js:37:12)
    at /data/sandcastle/boxes/instance-fbobjc-fbsource-catalyst-jest-macosx-local/fbobjc/Libraries/FBReactKit/js/react-native-github/node_modules/promise/lib/core.js:103:15
    at flush (/data/sandcastle/boxes/instance-fbobjc-fbsource-catalyst-jest-macosx-local/fbobjc/Libraries/FBReactKit/js/react-native-github/node_modules/promise/node_modules/asap/raw.js:50:29)
 line: undefined
 message: Requiring unknown module "InitializeJavaScriptAppEngine". If you are sure the module is there, try restarting the packager. [Error: Requiring unknown module "InitializeJavaScriptAppEngine". If you are sure the module is there, try restarting the packager.]

```

**Comment by skevy** [comment]
@vjeux ok yah these are new errors since the last time I rebased. Will rebase again and fix.

**Comment by vjeux** [comment]
@facebook-github-bot shipit

**Comment by skevy** [comment]
@vjeux that failed objc test in the travis report is unrelated to my code.

**Comment by vjeux** [comment]
Hmm, it is still failing because of `Cannot read property 'isNodeModulesDir' of undefined`

**Comment by skevy** [comment]
@vjeux it doesn't fail with that on my local or on Travis tho...https://travis-ci.org/facebook/react-native/jobs/93826649

Any idea what the difference in environment might be?

**Comment by vjeux** [comment]
Turns out, it doesn't work for our internal use case. We figured out another solution, instead of the `isNodeModule` and `providesModuleNodeModules`, make a `providesModulesWhitelist` array of paths (maybe a regex?) that whitelists places where @providesModule can appear.

```javascript
providesModuleWhitelist: [
  'RKJSModules/',
  'react-native-github/Libraries/',
  'react-native-github/node_modules/react-tools/'
]
```

**Comment by skevy** [comment]
Thanks @vjeux.

As we discussed, I'm going to close this PR and reopen a new one that has the correct fix.

**Comment by martinbigio** [comment]
@skevy I'm back :). Let me know when the PR is ready. Thanks for working on this!

**Comment by skevy** [comment]
@martinbigio yah I'm reopening this now, because we actually determined that the test failures were a Facebook internal infra problem (related to a babel-transform that isn't open sourced), not a problem with this PR.

Wait till @vjeux tells you about this one :)

**Comment by martinbigio** [comment]
Oh yeah, he told me how you guys fixed it. XP for the win :)

**Comment by martinbigio** [comment]
@cpojer do any of the modified file should be modified on node-haste as well?

File: packager/react-packager/src/DependencyResolver/Module.js (hunk 6 above)
Comment (martinbigio) [inline packager/react-packager/src/DependencyResolver/Module.js]: really like how concise this became :)

File: packager/react-packager/src/DependencyResolver/Module.js [hunk 7]
```diff
@@ -81,8 +82,11 @@ class Module {
     if (!this._reading) {
       this._reading = this._fastfs.readFile(this.path).then(content => {
         const data = {};
+
+        // Set an id on the module if it's using @providesModule syntax
         const moduleDocBlock = docblock.parseAsObject(content);
-        if (moduleDocBlock.providesModule || moduleDocBlock.provides) {
+        if (!this._depGraphHelpers.isNodeModulesDir(this.path) &&
```
Comment (martinbigio) [inline packager/react-packager/src/DependencyResolver/Module.js]: Maybe we should add a comment explaining why checking for this is necessary. It's not immediately obvious to me

**Comment by martinbigio** [comment]
Looks great to me. @cpojer what do you want to do with the changes that affect node-haste? How far away are we from getting rid of the duplicated code? Maybe we should think about a strategy to avoid missing changes.

File: packager/react-packager/src/DependencyResolver/Module.js (hunk 7 above)
Comment (skevy) [inline packager/react-packager/src/DependencyResolver/Module.js]: Sure np.

File: packager/react-packager/src/DependencyResolver/Module.js [hunk 8]
```diff
@@ -81,8 +82,16 @@ class Module {
     if (!this._reading) {
       this._reading = this._fastfs.readFile(this.path).then(content => {
         const data = {};
+
+        // Set an id on the module if it's using @providesModule syntax
+        // and if it's NOT in node_modules (and not a whitelisted node_module).
+        // This handles the case where a project may have a dep that has @providesModule
+        // docblock comments, but doesn't want it to conflict with whitelisted @providesModule
+        // modules, such as react-haste, fbjs-haste, or react-native or with non-dependency,
+        // project-specific code that is using @providesModule.
```
Comment (skevy) [inline packager/react-packager/src/DependencyResolver/Module.js]: @martinbigio check out this explanation and let me know if that's good.

**Comment by martinbigio** [comment]
I'll chat with @cpojer to see how we can pull changes into the new resolver repo

File: packager/react-packager/src/DependencyResolver/Module.js [hunk 9]
```diff
@@ -66,23 +67,31 @@ class Module {
...
-    return this._read().then(data => data.asyncDependencies);
+    return this.read().then(data => data.asyncDependencies);
   }
 
   invalidate() {
     this._cache.invalidate(this.path);
   }
 
-  _read() {
+  read() {
```
Comment (skevy) [inline packager/react-packager/src/DependencyResolver/Module.js]: @vjeux just pointing out that I renamed this method as we discussed, as to allow for proper class inheritance and to not break due to `babel-plugin-class-private-props`.

cc @martinbigio

**Comment by vjeux** [comment]
@facebook-github-bot shipit

**Comment by skevy** [comment]
Yay this finally merged! Thanks @vjeux
